analyzer/runtime/evm/erc165.go

```diff
@@ -0,0 +1,79 @@
+package evm
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/oasisprotocol/oasis-core/go/common/errors"
+
+	"github.com/oasisprotocol/oasis-indexer/analyzer/evmabi"
+	"github.com/oasisprotocol/oasis-indexer/log"
+	"github.com/oasisprotocol/oasis-indexer/storage/oasis/nodeapi"
+)
+
+func InterfaceID(abi *abi.ABI) []byte {
+	id := []byte{0, 0, 0, 0}
+	for _, m := range abi.Methods {
+		id[0] ^= m.ID[0]
+		id[1] ^= m.ID[1]
+		id[2] ^= m.ID[2]
+		id[3] ^= m.ID[3]
+	}
+	return id
+}
+
+var (
+	InvalidInterfaceID             = []byte{0xff, 0xff, 0xff, 0xff}
+	ERC165InterfaceID              = InterfaceID(evmabi.ERC165)
+	ERC721InterfaceID              = InterfaceID(evmabi.ERC721)
+	ERC721TokenReceiverInterfaceID = InterfaceID(evmabi.ERC721TokenReceiver)
+	ERC721MetadataInterfaceID      = InterfaceID(evmabi.ERC721Metadata)
+	ERC721EnumerableInterfaceID    = InterfaceID(evmabi.ERC721Enumerable)
+)
+
+const ERC165GasLimit uint64 = 30_000
+
+func callERC165SupportsInterface(ctx context.Context, source nodeapi.RuntimeApiLite, round uint64, contractEthAddr []byte, interfaceID []byte) (bool, error) {
+	var result bool
+	var interfaceIDFixed [4]byte
+	copy(interfaceIDFixed[:], interfaceID)
+	if err := evmCallWithABICustom(ctx, source, round, DefaultGasPrice, ERC165GasLimit, DefaultCaller, contractEthAddr, DefaultValue, evmabi.ERC165, &result, "supportsInterface", interfaceIDFixed); err != nil {
+		return false, err
+	}
+	return result, nil
+}
+
+func detectInterface(ctx context.Context, logger *log.Logger, source nodeapi.RuntimeApiLite, round uint64, contractEthAddr []byte, interfaceID []byte) (bool, bool, error) {
```

Comment:
Can you document the second return variable please?

It's weird to have `ok` and `err`. Can we fold the functionality of `ok` into the `err` somehow, perhaps by using `EVMDeterministicError`? (I haven't thought this through, just noting a weirdness.)

Reply:
ok on adding doc

about the design of this though, folding the functionality of `ok` into `err` I'm opposed to. the purpose of this is to handle a kind of error. handling is all of these things, and if we do something other than all or none of them, we'll harm the composability, break consistency, give programmers stomach problems, etc

1. make sure errors have a way to get to eyeballs -- detectInterface uses logDeterministicError
2. take us out of the idiomatic error workflow -- detectInterface returns nil err
3. substitute a consistent value/state/action/whatever -- detectInterface returns supports=indeterminate, ok=false

Reply:
oh you know what, it'll be simpler if I make a wrapper that &&'s the ok